tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,206 @@
+		teardown(releaseName)
+
+		cfg := getDefaultSetupCfg()
+		cfg.nfr = true
+
+		setup(cfg, "--set", "nginxGateway.securityContext.runAsNonRoot=false")
+
+		Expect(resourceManager.Apply([]client.Object{ns})).To(Succeed())
+		Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+		// Sometimes the traffic would error with code 502, after implementing this sleep it stopped.
+		time.Sleep(2 * time.Second)
+
+		expectWorkingTraffic(teaURL, coffeeURL)
+	})
+
+	AfterAll(func() {
+		Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.Delete([]client.Object{ns})).To(Succeed())
+	})
+
+	It("recovers when NGF container is restarted", func() {
+		podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(podNames).ToNot(BeEmpty())
+
+		output, err := restartNGFProcess()
+		Expect(err).ToNot(HaveOccurred(), string(output))
+		// Wait for NGF to restart.
+		time.Sleep(2 * time.Second)
+
+		expectWorkingTraffic(teaURL, coffeeURL)
+
+		checkContainerLogsForErrors(podNames[0])
+
+		// I tried just deleting the routes and ran into a bunch of issues, deleting all the files was better
+		Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+		// Wait for files to be deleted.
+		time.Sleep(2 * time.Second)
+
+		expectFailingTraffic(teaURL, coffeeURL)
+
+		Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+		expectWorkingTraffic(teaURL, coffeeURL)
+	})
+
+	It("recovers when nginx container is restarted", func() {
+		podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(podNames).ToNot(BeEmpty())
+
+		// If we correctly restart the nginx container, it should give us an error 137 code
+		output, err := restartNginxContainer()
+		Expect(err).To(HaveOccurred(), string(output))
+		time.Sleep(2 * time.Second)
+
+		checkContainerLogsForErrors(podNames[0])
+
+		Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+		// Wait for files to be deleted.
+		time.Sleep(2 * time.Second)
+
+		expectFailingTraffic(teaURL, coffeeURL)
+
+		Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+		expectWorkingTraffic(teaURL, coffeeURL)
+	})
+})
+
+func restartNginxContainer() ([]byte, error) {
+	podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(podNames).ToNot(BeEmpty())
+
+	// This should error since we are in the nginx container, killing the nginx master process.
+	// Exit Code 137 is the code we want to see as that is the kill -9 signal
```

Comment:
Could we not return error from this function if the exit code is 137?

I don't think it makes sense for the client of this function to know those details and thus check that this function returns an error in case of success.
How do we distinguish that error from any other error?

Same for restartNGFProcess

Reply:
Function no longer returns an error - realized it was user error as when I changed the script to use some regex on the pgrep command (`[]`) it no longer errored similar to the `restartNGFProcess`.

Added the restart count so now we can check that the command successfully restarted the container.